ddl/owner.go

```diff
@@ -0,0 +1,166 @@
+	etcdSession *concurrency.Session
+}
+
+func (w *worker) isOwner() bool {
+	return atomic.LoadInt32(&w.ddlOwner) == 1
+}
+
+func (w *worker) setOwner(isOwner bool) {
+	if isOwner {
+		atomic.StoreInt32(&w.ddlOwner, 1)
+	} else {
+		atomic.StoreInt32(&w.ddlOwner, 0)
+	}
+}
+
+func (w *worker) isBgOwner() bool {
+	return atomic.LoadInt32(&w.bgOwner) == 1
+}
+
+func (w *worker) setBgOwner(isOwner bool) {
+	if isOwner {
+		atomic.StoreInt32(&w.bgOwner, 1)
+	} else {
+		atomic.StoreInt32(&w.bgOwner, 0)
+	}
+}
+
+func (w *worker) newSession(ctx goctx.Context, retryCnt int) {
+	for i := 0; i < retryCnt; i++ {
+		session, err := concurrency.NewSession(w.etcdClient, concurrency.WithContext(ctx))
+		if err != nil {
+			log.Warnf("[ddl] failed to new session, err %v", err)
+			time.Sleep(200 * time.Millisecond)
+			continue
+		}
+		w.etcdSession = session
+		break
+	}
+}
+
+func (d *ddl) campaignOwners(ctx goctx.Context) {
+	d.worker.newSession(ctx, newSessionDefaultRetryCnt)
+	go d.campaignLoop(ctx, ddlOwnerKey)
+	d.campaignLoop(ctx, bgOwnerKey)
+}
+
+func (d *ddl) campaignLoop(ctx goctx.Context, key string) {
+	defer d.wait.Done()
+	worker := d.worker
+	for {
+		if d.isClosed() {
+			return
+		}
+
+		elec := concurrency.NewElection(worker.etcdSession, key)
+		err := elec.Campaign(ctx, worker.ddlID)
+		if err != nil {
+			log.Infof("[ddl] worker %s failed to campaign, err %v", worker.ddlID, err)
+			continue
+		}
+
+		// Get owner information.
+		resp, err := elec.Leader(ctx)
+		if err != nil {
+			// If no leader elected currently, it returns ErrElectionNoLeader.
+			log.Infof("[ddl] failed to get leader, err %v", err)
+			continue
+		}
+		leader := string(resp.Kvs[0].Value)
+		log.Info("[ddl] %s worker is %s, owner is %v", key, worker.ddlID, leader)
+		if leader == worker.ddlID {
+			worker.setOwnerVal(key, true)
+		} else {
+			log.Warnf("[ddl] worker %s doesn't the owner", worker.ddlID)
+			continue
+		}
+
+		// TODO: Use content instead of quitCh.
+		worker.watchOwner(ctx, string(resp.Kvs[0].Key))
+		worker.setOwnerVal(key, false)
+		d.hookMu.Lock()
+		d.hook.OnWatched(ctx)
+		d.hookMu.Unlock()
+
+		select {
+		case <-worker.etcdSession.Done():
+			// TODO: Create session again?
+		default:
+		}
+	}
+}
+
+func (w *worker) setOwnerVal(key string, val bool) {
+	if key == ddlOwnerKey {
+		w.setOwner(val)
+	} else {
+		w.setBgOwner(val)
+	}
+}
+
+func (w *worker) watchOwner(ctx goctx.Context, key string) {
+	log.Debugf("[ddl] worker %s watch owner key %v", w.ddlID, key)
+	watchCh := w.etcdClient.Watch(ctx, key)
+	for {
+		select {
+		case resp := <-watchCh:
+			if resp.Canceled {
+				log.Infof("[ddl] worker %s watch owner key %v failed, no owner",
+					w.ddlID, key)
+				return
+			}
+
+			for _, ev := range resp.Events {
+				if ev.Type == mvccpb.DELETE {
+					log.Infof("[ddl] worker %s watch owner key %v failed, owner is deleted", w.ddlID, key)
+					return
+				}
+			}
+		case <-w.etcdSession.Done():
+			// TODO: Create session again?
```

Comment:
When session is `Done()`, this worker is not owner any more, this function should return and campaign leader...